Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp

```diff
@@ -759,9 +784,19 @@ JSValue WebAssemblyModuleRecord::evaluate(JSGlobalObject* globalObject)
         for (const Wasm::Segment::Ptr& segment : data) {
             if (!segment->isActive())
                 continue;
-            uint32_t offset = segment->offsetIfActive->isGlobalImport()
-                ? static_cast<uint32_t>(m_instance->instance().loadI32Global(segment->offsetIfActive->globalImportIndex()))
-                : segment->offsetIfActive->constValue();
+            uint32_t offset = 0;
+            if (segment->offsetIfActive->isGlobalImport())
+                offset = static_cast<uint32_t>(m_instance->instance().loadI32Global(segment->offsetIfActive->globalImportIndex()));
+            else if (segment->offsetIfActive->isConst())
+                offset = segment->offsetIfActive->constValue();
+            else {
+                uint64_t result;
+                if (!evaluateConstantExpression(moduleInformation.constantExpressions[segment->offsetIfActive->constantExpressionIndex()], moduleInformation, Wasm::Types::I32, result)) {
+                    exception = JSValue(throwException(globalObject, scope, createJSWebAssemblyRuntimeError(globalObject, vm, "couldn't evaluate data offset constant expression"_s)));
```

Comment:
Can we get a more descriptive error?

Reply:
I've changed this now to append the underlying error message from the constant expression generator, and abstracted it a bit so the error is created in the helper function.